### PR TITLE
Add support for custom fields

### PR DESCRIPTION
**Pull request by hisabimbola**
The biggest change is adding support for custom fields in the inventory. Rewrite the transformation function to be promise based because of async fetching the correct custom field type from the API. When the same type of custom field is being used multiple times it is only fetched once using memorisation.

Secondly the stock fields now have the same name as the API, "quantity" > "quantityOnStock" because header fields are no longer ignored but used to map the correct values. *- Selwyn* :)

### CODE REVIEW

**Comment by Siilwyn** [comment]
PR ready to be reviewed @emmenko @PhilippSpo

File: README.md [hunk 1]
```diff
@@ -30,16 +30,20 @@ When using SFTP, you should not use the `--file` option, instead you need to pro
...
 
-Column 1 will be used as `SKU` identifier, whereas column 2 will be used as `quantity`.
-An example:
+A simple example:
 ```
-sku,quantity
-foo,9
-bar,-1
-SKU-123,42
+sku,quantityOnStock,restockableInDays,supplyChannel,expectedDelivery
```
Comment (emmenko) [inline README.md]: FYI: since we're introducing a breaking change with `quantityOnStock`, we need to bump a major version + document this in the release notes.

We can also try to make it backwards compatible:
- `quantityOnStock` is the new default
- if client uses `quantity`, we map it internally to `quantityOnStock` + we warn the client that this is deprecated

File: src/coffee/constants.coffee [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  constants.HEADER_ID,
+  constants.HEADER_SKU,
+  constants.HEADER_PRICES,
+  constants.HEADER_TAX,
+  constants.HEADER_CATEGORIES,
+  constants.HEADER_IMAGES,
+  # TODO: image labels and dimensions
+]
+
+exports.ALL_HEADERS = exports.BASE_HEADERS.concat(exports.BASE_LOCALIZED_HEADERS.concat(exports.SPECIAL_HEADERS))
```
Comment (emmenko) [inline src/coffee/constants.coffee]: Hmm, this file looks copy-pasted from the product-csv-sync. Can you check that we need all those constants?

File: src/coffee/mappings.coffee [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+_.mixin require('underscore-mixins')
+csv = require 'csv'
+CONS = require './constants'
+
+class CustomFieldMappings
+
+  constructor: (options = {}) ->
+    @errors = []
+
+  mapFieldTypes: (fieldDefinitions, typeDefinitionKey, rowIndex, key, value,langHeader) ->
```
Comment (emmenko) [inline src/coffee/mappings.coffee]: Nitpick: usually if we start having more then 2 arguments, I would suggest you use named arguments, makes it easy to refactor things.

```coffeescript
mapFieldTypes: ({ fieldDefinitions, typeDefinitionKey, rowIndex, key, value, langHeader }) ->
```

File: src/coffee/stockimport.coffee [hunk 4]
```diff
@@ -90,7 +89,7 @@ class StockImport
         ElasticIo.returnFailure err, next
       .done()
     else
-      ElasticIo.returnFailure "#{LOG_PREFIX}No data found in elastic.io msg.", next
+      ElasticIo.returnFailure "#{CONS.LOG_PREFIX}No data found in elastic.io msg.", next
```
Comment (emmenko) [inline src/coffee/stockimport.coffee]: FYI: to be honest, everything related to `elastic.io` can be removed. Check with @hajoeichler first please

File: src/coffee/stockimport.coffee [hunk 5]
```diff
@@ -130,27 +129,26 @@ class StockImport
...
+                .then (result) -> resolve result
+            .catch (err) -> reject err
+            .done()
 
   performStream: (chunk, cb) ->
     @_processBatches(chunk).then -> cb()
 
   _getHeaderIndexes: (headers, csvHeaders) ->
-    Promise.all _.map csvHeaders.split(','), (h) =>
+    Promise.all _.map csvHeaders.split(','), (h) ->
```
Comment (emmenko) [inline src/coffee/stockimport.coffee]: Question: at this point, the parsed `csvHeaders` are always comma-separated or should we use the given `csvDelimiter` ?

File: src/coffee/stockimport.coffee [hunk 6]
```diff
@@ -179,19 +177,94 @@ class StockImport
...
+            Promise.resolve(@_mapCellData(cell, headerName)).then (cellData) =>
+              _data[headerName] = cellData
+
+        ).then =>
+          if _.size(@customFieldMappings.errors) isnt 0
+            return cb @customFieldMappings.errors
+          cb null, _data
+      , (err, data) ->
+        if err
+          reject(err)
```
Comment (emmenko) [inline src/coffee/stockimport.coffee]: Probably best to use `return` here, or use `if-else`

File: src/coffee/stockimport.coffee [hunk 7]
```diff
@@ -179,19 +177,94 @@ class StockImport
...
+        @customFieldMappings.mapFieldTypes customTypeDefinition.fieldDefinitions,customTypeDefinition.key,rowIndex,fieldName,cell,lang
+    else
+      data.custom.fields[fieldName] = @customFieldMappings.mapFieldTypes customTypeDefinition.fieldDefinitions,customTypeDefinition.key,rowIndex,fieldName,cell
+
+
+  _getCustomTypeDefinition: _.memoize (customTypeKey) ->
+    @__getCustomTypeDefinition customTypeKey
+
+  # Should not be called directly.
+  __getCustomTypeDefinition: (customTypeKey) ->
```
Comment (emmenko) [inline src/coffee/stockimport.coffee]: Just curious, why are you using `__` instead of `_` ?

File: src/coffee/stockimport.coffee [hunk 8]
```diff
@@ -179,19 +177,94 @@ class StockImport
...
+    if lang
+      data.custom.fields[fieldName] =
+        _.defaults (data.custom.fields[fieldName] || {}),
+        @customFieldMappings.mapFieldTypes customTypeDefinition.fieldDefinitions,customTypeDefinition.key,rowIndex,fieldName,cell,lang
+    else
+      data.custom.fields[fieldName] = @customFieldMappings.mapFieldTypes customTypeDefinition.fieldDefinitions,customTypeDefinition.key,rowIndex,fieldName,cell
+
+
+  _getCustomTypeDefinition: _.memoize (customTypeKey) ->
+    @__getCustomTypeDefinition customTypeKey
```
Comment (emmenko) [inline src/coffee/stockimport.coffee]: Can you explain how `memoize` help in this case? Not just for me but in the code also, for future references ;)

File: src/coffee/run.coffee [hunk 9]
```diff
@@ -35,7 +35,7 @@ argv = require('optimist')
   .describe('logDir', 'directory to store logs')
   .describe('logSilent', 'use console to print messages')
   .describe('timeout', 'Set timeout for requests')
-  .default('csvHeaders', 'sku, quantity')
+  .default('csvHeaders', 'sku, quantityOnStock')
```
Comment (emmenko) [inline src/coffee/run.coffee]: Btw: this option is used to pass custom header fields. Are we still going to support this or the headers will be fixed?

File: README.md (hunk 1 above)
Comment (Siilwyn) [inline README.md]: Yes definitely a major bump, going to go with the second option: showing a deprecation warning.

File: src/coffee/stockimport.coffee (hunk 4 above)
Comment (Siilwyn) [inline src/coffee/stockimport.coffee]: Whoops, I hope not, it took me some time to get everything working with `elastic.io` after the changes...

File: src/coffee/stockimport.coffee (hunk 5 above)
Comment (Siilwyn) [inline src/coffee/stockimport.coffee]: We should use the `csvDelimiter`.

File: src/coffee/stockimport.coffee (hunk 6 above)
Comment (Siilwyn) [inline src/coffee/stockimport.coffee]: Definitely, just fixed this.

File: src/coffee/run.coffee (hunk 9 above)
Comment (hisabimbola) [inline src/coffee/run.coffee]: No, we are not supporting custom headers again.
We using the API key names as a guide.
That removes some level of complexity and mappings

File: README.md [hunk 10]
```diff
@@ -30,16 +30,20 @@ When using SFTP, you should not use the `--file` option, instead you need to pro
...
+bar,-1,3,channel-key,2016-10-27T14:36:04.487Z
+SKU-123,42,3,other-channel,2016-10-27T14:36:04.487Z
 ```
 
-> Please note that the header names are currently ignored.
+### Custom fields
+```
+sku,quantityOnStock,customType,customField.foo,customField.bar
+123,77,my-type,12,nac
+abc,-3,my-type,5,ho
```
Comment (PhilippSpo) [inline README.md]: Can one import different custom types at once?
```
sku,quantityOnStock,customType,customField.foo,customField.bar
123,77,my-type,12,nac
abc,-3,my-type-2,,ho
```
Should the fields that a custom type does not allow be just left empty like in my example?

Comment (hisabimbola) [inline README.md]: Yeah, it's possible to have more than one customType in one csv sheet and yes if you leave the field empty it will ignore it.

I will add a test for this use case

File: src/coffee/stockimport.coffee (hunk 8 above)
Comment (Siilwyn) [inline src/coffee/stockimport.coffee]: It helps because when having multiple inventory entries with the same customType we won't make the same call to the API to get the definition instead we get it from the 'cache' so to say.

**Comment by hajoeichler** [comment]
Everything related to `elastic.io` can be removed within the code.

File: src/coffee/stockimport.coffee (hunk 4 above)
Comment (Siilwyn) [inline src/coffee/stockimport.coffee]: Will be removed in a separate PR.

**Comment by Siilwyn** [comment]
@emmenko @PhilippSpo please take a final look. 🔬
